### PR TITLE
[TorchFX]: Test Fast Bias Correction algorithm

### PR DESCRIPTION
**Pull request by anzr299**
### Changes

Add Fast BC test in tests/torch/fx. 

### Related tickets

#2809 

### Tests

The implemented test class inherits from the template for Fast bias correction.

### CODE REVIEW

**Comment by anzr299** [comment]
precommit/common fails probably due to #2814

File: tests/torch/fx/test_fast_bias_correction.py [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+from nncf.quantization.algorithms.fast_bias_correction.torch_fx_backend import FXFastBiasCorrectionAlgoBackend
+from nncf.torch.dynamic_graph.patch_pytorch import disable_patching
+from nncf.torch.nncf_network import NNCFNetwork
+from tests.post_training.test_templates.test_fast_bias_correction import TemplateTestFBCAlgorithm
+from tests.torch.fx.graph_manager import get_bias_value
+from tests.torch.fx.graph_manager import is_node_with_bias
+
+
+class TestTorchFXFBCAlgorithm(TemplateTestFBCAlgorithm):
+    @staticmethod
```
Comment (daniil-lyakhov) [inline tests/torch/fx/test_fast_bias_correction.py:29]: Please implement a base class as it done in torch tests

**Comment by daniil-lyakhov** [comment]
@anzr299, Nice work!

File: tests/torch/fx/test_fast_bias_correction.py (hunk 1 above)
Comment (anzr299) [reply tests/torch/fx/test_fast_bias_correction.py:29]: I'm sorry I didnt really get the suggestion, the template class is already implemented right?

Comment (daniil-lyakhov) [reply tests/torch/fx/test_fast_bias_correction.py:29]: Yes, I was wrong. Please inherit `TestTorchFXFBCAlgorithm` in `TestTorchFXCudaFBCAlgorithm` and keep common methods (like get_backend_model and etc) in `TestTorchFXFBCAlgorithm`

Comment (anzr299) [reply tests/torch/fx/test_fast_bias_correction.py:29]: Oh, my mistake there, I will get those changes back and refactor the code.

Comment (anzr299) [reply tests/torch/fx/test_fast_bias_correction.py:26]: Done!

File: tests/torch/fx/graph_manager.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+
+import nncf
+from nncf.common.factory import NNCFGraphFactory
+from nncf.common.graph.graph import NNCFGraph
+from nncf.common.graph.graph import NNCFNode
+from nncf.torch.model_graph_manager import OPERATORS_WITH_BIAS_METATYPES
+from nncf.torch.model_graph_manager import find_const_node_in_constant_subgraph
+
+
+def is_node_with_bias(node: NNCFNode, nncf_graph: NNCFGraph) -> bool:
```
Comment (alexsu52) [inline tests/torch/fx/graph_manager.py:25]: @anzr299, @daniil-lyakhov, it looks like, this is general functions and can be used in the fast bias correction. Please consider possible refactoring to use functions in the NNCF algorithms.

**Review by alexsu52 — CHANGES_REQUESTED** [review]

File: tests/torch/fx/graph_manager.py (hunk 2 above)
Comment (anzr299) [reply tests/torch/fx/graph_manager.py]: I have removed it for this PR for now. We can add it back in a follow up PR when refactoring

File: nncf/experimental/torch/fx/helpers.py
[the remark this answers is not in the document]
Comment (anzr299) [reply nncf/experimental/torch/fx/helpers.py:1]: Alright, I removed the helpers.py file now

**Comment by daniil-lyakhov** [comment]
@anzr299, please rebase your changes

**Comment by anzr299** [comment]
> @anzr299, please rebase your changes

Done

File: nncf/experimental/torch/fx/transformations.py [hunk 3]
```diff
@@ -243,6 +243,63 @@ def _set_module_to_the_graph_module(
...
+
+
+def is_conv_node(node: torch.fx.Node):
+    """
+    Return whether the node refers to an aten conv op.
+    """
+    return node.op == "call_function" and node.target in [
+        torch.ops.aten.conv1d.default,
+        torch.ops.aten.conv2d.default,
+    ]
```
Comment (daniil-lyakhov) [inline nncf/experimental/torch/fx/transformations.py:265]: Please use `_is_conv` function

File: nncf/experimental/torch/fx/transformations.py [hunk 4]
```diff
@@ -243,6 +243,63 @@ def _set_module_to_the_graph_module(
...
+
+
+def is_bn_node(node: torch.fx.Node):
```
Comment (daniil-lyakhov) [inline nncf/experimental/torch/fx/transformations.py:268]:
```suggestion
def _is_bn_node(node: torch.fx.Node):
```

File: nncf/experimental/torch/fx/transformations.py [hunk 5]
```diff
@@ -243,6 +243,63 @@ def _set_module_to_the_graph_module(
     return module_name_in_model
 
 
+def is_supported_batch_norm_for_training(node: torch.fx.Node):
```
Comment (daniil-lyakhov) [inline nncf/experimental/torch/fx/transformations.py:246]:
```suggestion
def _is_supported_batch_norm_for_training(node: torch.fx.Node):
```

File: tests/torch/fx/test_fast_bias_correction.py [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+    def list_to_backend_type(data: List) -> torch.Tensor:
+        return torch.Tensor(data)
+
+    @staticmethod
+    def get_backend() -> FXFastBiasCorrectionAlgoBackend:
+        return FXFastBiasCorrectionAlgoBackend
+
+    @staticmethod
+    def backend_specific_model(model: torch.nn.Module, tmp_dir: str):
+        fx_model = get_fx_model(model)
```
Comment (daniil-lyakhov) [inline tests/torch/fx/test_fast_bias_correction.py:35]: Please move `get_fx_model` function here and use `super().get_fx_model()` in `TestTorchCudaFBCAlgorithm.backend_specific_model`

File: tests/torch/fx/test_fast_bias_correction.py [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+        return fx_cuda_model
+
+    @staticmethod
+    def fn_to_type(tensor):
+        return torch.Tensor(tensor).cuda()
+
+    @staticmethod
+    def check_bias(model: torch.fx.GraphModule, ref_bias: list):
+
+        TestTorchFXFBCAlgorithm.check_bias(model, ref_bias)
```
Comment (daniil-lyakhov) [inline tests/torch/fx/test_fast_bias_correction.py:84]:
```suggestion
```

**Review by daniil-lyakhov — CHANGES_REQUESTED** [review]

File: tests/torch/fx/test_fast_bias_correction.py (hunk 6 above)
Comment (anzr299) [reply tests/torch/fx/test_fast_bias_correction.py]: Done

File: tests/torch/fx/test_fast_bias_correction.py (hunk 7 above)
Comment (anzr299) [reply tests/torch/fx/test_fast_bias_correction.py]: Done

File: nncf/experimental/torch/fx/transformations.py (hunk 5 above)
Comment (anzr299) [reply nncf/experimental/torch/fx/transformations.py]: Done

File: nncf/experimental/torch/fx/transformations.py (hunk 4 above)
Comment (anzr299) [reply nncf/experimental/torch/fx/transformations.py]: Done

File: nncf/experimental/torch/fx/transformations.py (hunk 3 above)
Comment (anzr299) [reply nncf/experimental/torch/fx/transformations.py]: Done

File: tests/torch/fx/test_fast_bias_correction.py [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+        ex_input = torch.ones(input_shape).to(device)
+        model.eval()
+        with disable_patching():
+            exported_model = capture_pre_autograd_graph(model, args=(ex_input,))
+        apply_quantization_transformations(exported_model)
+        return exported_model
+
+    @staticmethod
+    def backend_specific_model(model: torch.nn.Module, tmp_dir: str):
+        fx_model = TestTorchFXFBCAlgorithm._get_fx_model(model)
```
Comment (daniil-lyakhov) [inline tests/torch/fx/test_fast_bias_correction.py:49]: Please move code from `_get_fx_model` to `backend_specific_model`

File: tests/torch/fx/test_fast_bias_correction.py [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+@pytest.mark.cuda
+@pytest.mark.skipif(not torch.cuda.is_available(), reason="Skipping for CPU-only setups")
+class TestTorchCudaFBCAlgorithm(TestTorchFXFBCAlgorithm):
+    @staticmethod
+    def list_to_backend_type(data: List) -> torch.Tensor:
+        return torch.Tensor(data).cuda()
+
+    @staticmethod
+    def backend_specific_model(model: bool, tmp_dir: str):
+        fx_cuda_model = super(TestTorchCudaFBCAlgorithm, TestTorchCudaFBCAlgorithm)._get_fx_model(model.cuda())
```
Comment (daniil-lyakhov) [inline tests/torch/fx/test_fast_bias_correction.py:88]:
```suggestion
        fx_cuda_model = super()._get_fx_model(model.cuda())
```

File: nncf/experimental/torch/fx/transformations.py (hunk 3 above)
Comment (daniil-lyakhov) [reply nncf/experimental/torch/fx/transformations.py]: There are two `_is_conv_node` function in the file, please remove one of the functions

Comment (anzr299) [reply nncf/experimental/torch/fx/transformations.py]: Oh, I misunderstood the suggestion. Sorry about that.

File: tests/torch/fx/test_fast_bias_correction.py (hunk 9 above)
Comment (anzr299) [reply tests/torch/fx/test_fast_bias_correction.py:88]: using super() causes problems due to the staticmethod i believe

Comment (daniil-lyakhov) [reply tests/torch/fx/test_fast_bias_correction.py:88]: You are right, please do `TestTorchFBCAlgorith.backend_specific_model(model.cuda(), tmp_dir)`

Comment (anzr299) [reply tests/torch/fx/test_fast_bias_correction.py:88]: Alright

Comment (anzr299) [reply tests/torch/fx/test_fast_bias_correction.py:88]: Will that not be bad for the class hirearchy in case we ever add more classes in between?

Comment (daniil-lyakhov) [reply tests/torch/fx/test_fast_bias_correction.py:88]: Ok, let's keep it as it is, but please remove arguments from `super` call

Comment (anzr299) [reply tests/torch/fx/test_fast_bias_correction.py:88]: But using `super` causes problems. Would you suggest I go ahead with `TestTorchFBCAlgorith.backend_specific_model(model.cuda(), tmp_dir)` anyways?

Comment (daniil-lyakhov) [reply tests/torch/fx/test_fast_bias_correction.py:88]: Sorry for unclear comments above, think best choice for now is to move this function outside of the function (but left in the same file, `test_bias_correction_aglorihtm` and call it in both `backend_specific_model` methods with different params.

Comment (anzr299) [reply tests/torch/fx/test_fast_bias_correction.py]: Done

File: tests/torch/fx/test_fast_bias_correction.py (hunk 8 above)
Comment (anzr299) [reply tests/torch/fx/test_fast_bias_correction.py:50]: Done

**Review by daniil-lyakhov — APPROVED** [review]

**Review by alexsu52 — APPROVED** [review]
LGTM

**Pull request merged**